lib/dal.ex

```diff
@@ -329,17 +332,27 @@ defmodule DAL do
   @doc """
   Convenience function for using repo discovery.
   """
-  @spec discover(queryable :: Ecto.Query.t() | Ecto.Changeset.t()) :: Ecto.Repo.t()
+  @spec discover(
+          queryable ::
+            nil | Ecto.Query.t() | Ecto.Changeset.t() | Ecto.Schema.t() | [Ecto.Schema.t()]
+        ) :: Ecto.Repo.t()
   def discover(%Ecto.Changeset{data: data}) do
-    Discovery.fetch(data)
+    discover(data)
+  end
+
+  def discover(structs) when is_list(structs) do
+    structs |> Enum.find(& &1) |> discover()
   end
 
+  def discover(nil), do: nil
+
   def discover(queryable) do
     Discovery.fetch(queryable)
   end
 
   def preload(struct_or_structs_or_nil, preloads, opts \\ []) do
-    Ecto.Repo.Preloader.preload(struct_or_structs_or_nil, DAL, preloads, opts)
+    repo = discover(struct_or_structs_or_nil)
+    Ecto.Repo.Preloader.preload(struct_or_structs_or_nil, repo, preloads, opts)
```

Comment:
What's the motivation for this change?

Reply:
So we were passing in `DAL` to `Ecto.Repo.Preloader.preload/4` and in Ecto 2, it was calling `.all` on the input argument. But in Ecto 3, the implementation has changed. Now it expects this second argument to be an actual repo rather than a module name and it calls 
```
```

Reply:
OK understood. The `DAL` should look like a Repo anyway but I think it's inconsequential.